core/src/test/java/kafka/log/remote/RemoteLogManagerInteractionTest.java

```diff
@@ -0,0 +1,323 @@
+        RemoteLogManager.RLMTask task = remoteLogManager.new RLMCopyTask(topicIdPartition, 128);
+
+        buildSegment(retentionSize);
+        buildSegment(retentionSize);
+        buildSegment(retentionSize);
+        log.maybeUpdateHighWatermark(log.logEndOffset());
+
+        task.run();
+
+        verify(remoteStorageManager, times(3)).copyLogSegmentData(any(), any());
+    }
+
+    @Test
+    void failedSegmentAreRetried() throws RemoteStorageException {
+        RemoteLogManager.RLMTask task = remoteLogManager.new RLMCopyTask(topicIdPartition, 128);
+
+        buildSegment(retentionSize);
+        buildSegment(retentionSize);
+        buildSegment(retentionSize);
+        log.maybeUpdateHighWatermark(log.logEndOffset());
+
+        when(remoteStorageManager.copyLogSegmentData(any(), any()))
+            .thenThrow(new RemoteStorageException(""))
+            .thenReturn(Optional.empty());
+
+        task.run();
+        task.run();
+
+        verify(remoteStorageManager, times(4)).copyLogSegmentData(any(), any());
+    }
+
+    @Test
+    void failedSegmentsAreNotAccountedInSizeForRetention() throws RemoteStorageException {
+        // We're testing the scenario where the first segment is correctly pushed then the next segment fails to get
+        // uploaded multiple times before succeeding again.
+        // At the end we should see no call to delete as retention.bytes was never breached
+        RemoteLogManager.RLMTask copyTask = remoteLogManager.new RLMCopyTask(topicIdPartition, 128);
+        RemoteLogManager.RLMTask cleanupTask = remoteLogManager.new RLMExpirationTask(topicIdPartition);
+
+        buildSegment(retentionSize /  3);
+        log.maybeUpdateHighWatermark(log.logEndOffset());
+        when(remoteStorageManager.copyLogSegmentData(any(), any())).thenReturn(Optional.empty());
+        copyTask.run();
+
+        buildSegment(retentionSize /  3);
+        log.maybeUpdateHighWatermark(log.logEndOffset());
+        when(remoteStorageManager.copyLogSegmentData(any(), any())).thenThrow(new RemoteStorageException(""));
+        for (int i = 0; i < 4; i++) {
+            copyTask.run();
+        }
+        doReturn(Optional.empty()).when(remoteStorageManager).copyLogSegmentData(any(), any());
+        copyTask.run();
+
+        buildSegment(retentionSize /  3);
+        log.maybeUpdateHighWatermark(log.logEndOffset());
+        copyTask.run();
+        cleanupTask.run();
+
+        // total is 1 (success) + 4 (failures) + 2x1 (success)
+        verify(remoteStorageManager, times(7)).copyLogSegmentData(any(), any());
+        verify(remoteStorageManager, never()).deleteLogSegmentData(
+            argThat(segment -> segment.state().equals(RemoteLogSegmentState.COPY_SEGMENT_FINISHED))
+        );
+        verify(remoteStorageManager, times(4)).deleteLogSegmentData(
+            argThat(segment -> segment.state().equals(RemoteLogSegmentState.COPY_SEGMENT_STARTED))
+        );
```

Comment:
Could you help me understand why we call 4 times of `deleteLogSegmentData` for `COPY_SEGMENT_STARTED` here? Maybe add the explanation as a comment above the line.

Reply:
I'll add this as a comment but the idea is that, because we have a `COPY_SEGMENT_FINISHED` after a `COPY_SEGMENT_STARTED` in the metadata for this topic partition, all the `COPY_SEGMENT_STARTED` segments will be seen as dangling and will be deleted.